### DISCUSSION TITLE
How to read a byte array with MagickFormat.Gray?

### DISCUSSION THREAD

**Discussion by kmgallahan** [dlemstra/Magick.NET#1451] [Help]
I'm trying to create a MagickImage from a byte array that is of the format `MagickFormat.Gray`.

I always get this error:

> ImageMagick.MagickOptionErrorException: must specify image size `' @ error/gray.c/ReadGRAYImage/147

It's unclear to me how, what, or where the size is supposed to come from. 

I've tried many variations of this pseudo-code without luck:

``` cs
        using var image = new MagickImage(@"C:\Dev\Tests\SampleImages\01HBXXRH4514PNVVVYKYZ0M6KN.tif");
        using var singleChannelImage = image.Clone();
        singleChannelImage.ColorType = ColorType.Grayscale;

        var byteArray = singleChannelImage.ToByteArray(MagickFormat.Gray);

        var newImage = new MagickImage(byteArray, MagickFormat.Gray);

        // or variations of singleChannelImage.Read(byteArray, ....
```

If I exclude MagickFormat.Gray on the create/read operation, then I get this:

> ImageMagick.MagickMissingDelegateErrorException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/460

Best I've found so far is:

``` cs
        singleChannelImage.ReadPixels(
            byteArray, 
            new PixelReadSettings(imageSize.Width, imageSize.Height, StorageType.Quantum, "R"));

            // Unsure how "correct/valid" that mapping string is
```

---

Additional context: just trying to read an image, make it grayscale (single channel image), pass it elsewhere for processing as a byte array, get back the results, and read it in.

I started with a `Span<byte>`s but that wasn't working either, so tried out the array.

**Comment by dlemstra** [comment]
You should use the `ReadPixels` method for this. Did that not work for you?

**Reply by kmgallahan** [reply]
Yes, I did eventually came to use `ReadPixels` with the parameters shown above. However, I struggled a bit due to:

1. The Magick.NET `MagickImage` ctors take any `MagicFormat` without validation/complaining
2. ImageMagick throws the error shown indicating that size needs to (and can) be specified
3. Here is where I spent time trying to figure out how to pass in the size, only to eventually realize it probably just isn't an option

I do realize these things:

* Checking for invalid `MagicFormat` input would add a performance hit for 99.999% of usage (maybe trivial though)
* Passing the current image size to ImageMagick would require special case selection
* Adding ctors that take size params bloats the API
* Documenting valid format selections in the xml could help, but can easily be overlooked

So, I guess... 🤷‍♂️

**Reply by dlemstra** [reply]
> 1. The Magick.NET MagickImage ctors take any MagicFormat without validation/complaining

You were getting exceptions when you did not provide the correct byte array for the `Gray` format? This is valid in the constructor but not for your specific situation. But maybe I could add some extra documentation to the `byte[]` overloads that if you want to read pixels you should use the `ReadPixels` method instead.

> 2. ImageMagick throws the error shown indicating that size needs to (and can) be specified

This exception was thrown because the `Gray` format requires you to specify the dimensions of the image. 

> 3. Here is where I spent time trying to figure out how to pass in the size, only to eventually realize it probably just isn't an option

This is a fair point but that is probably because the message from ImageMagick is thrown as an exception. But I don't want to change them because they are thrown at different spots and that would probably require some regex magic.

**Reply by kmgallahan** [reply]
> You were getting exceptions when you did not provide the correct byte array for the Gray format?

I produced the byte array in Gray format from Magick.NET, used elsewhere, and then needed to ingest back into Magick.NET.

* `ReadonlySpan<byte>` ctor failed with size error, couldn't find size param
* `MagickImage.Read` `ReadonlySpan<byte>` same error, couldn't find size param and current size wasn't used
* Same results trying to use byte array versions
* Posted this question
* Realized `ReadPixels` could be used and edited the question

I imagine trying to use any of the 'raw sample' formats in the byte array or span ctors or Read methods would produce the same result (getting a size-related error message):

- BAYER
- CMYK
- CMYKA
- GRAY
- GRAYA
- RAW
- RGB
- RGB565
- RGBA
- YCbCr
- YCbCrA

The most helpful would be to check the format first and throw an error indicating that these formats require the use of `ReadPixels`, rather than passing anything to ImageMagick.

Not sure if any other input validation is done in Magick.NET though, so this might be something you're trying to avoid.
